tsl/src/compression/compression_storage.c

```diff
@@ -0,0 +1,389 @@
+	Datum toast_options;
+	static char *validnsps[] = HEAP_RELOPT_NAMESPACES;
+	CompressionSettings *settings = ts_compression_settings_get(src_chunk->hypertable_relid);
+
+	Oid owner = ts_rel_get_owner(chunk->hypertable_relid);
+
+	CreateStmt *create;
+	RangeVar *compress_rel;
+
+	create = makeNode(CreateStmt);
+	create->tableElts = column_defs;
+	create->inhRelations = NIL;
+	create->ofTypename = NULL;
+	create->constraints = NIL;
+	create->options = NULL;
+	create->oncommit = ONCOMMIT_NOOP;
+	create->tablespacename = get_tablespace_name(tablespace_oid);
+	create->if_not_exists = false;
+
+	/* Invalid tablespace_oid <=> NULL tablespace name */
+	Assert(!OidIsValid(tablespace_oid) == (create->tablespacename == NULL));
+
+	/* create the compression table */
+	/* NewRelationCreateToastTable calls CommandCounterIncrement */
+	ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+	compress_rel = makeRangeVar(NameStr(chunk->fd.schema_name), NameStr(chunk->fd.table_name), -1);
+
+	create->relation = compress_rel;
+	tbladdress = DefineRelation(create, RELKIND_RELATION, owner, NULL, NULL);
+	CommandCounterIncrement();
+	chunk->table_id = tbladdress.objectId;
+	ts_copy_relation_acl(chunk->hypertable_relid, chunk->table_id, owner);
+	toast_options =
+		transformRelOptions((Datum) 0, create->options, "toast", validnsps, true, false);
+	(void) heap_reloptions(RELKIND_TOASTVALUE, toast_options, true);
+	NewRelationCreateToastTable(chunk->table_id, toast_options);
+	ts_catalog_restore_user(&sec_ctx);
+	modify_compressed_toast_table_storage(settings, column_defs, chunk->table_id);
+
+	set_statistics_on_compressed_chunk(chunk->table_id);
+	set_toast_tuple_target_on_chunk(chunk->table_id);
+
+	create_compressed_chunk_indexes(chunk, settings);
+
+	clone_constraints_to_chunk(src_chunk->hypertable_relid, chunk);
+
+	return chunk->table_id;
+}
+
+static void
+set_toast_tuple_target_on_chunk(Oid compressed_table_id)
+{
+	DefElem def_elem = {
+		.type = T_DefElem,
+		.defname = "toast_tuple_target",
+		.arg = (Node *) makeInteger(128),
+		.defaction = DEFELEM_SET,
+		.location = -1,
+	};
+	AlterTableCmd cmd = {
+		.type = T_AlterTableCmd,
+		.subtype = AT_SetRelOptions,
+		.def = (Node *) list_make1(&def_elem),
+	};
+	ts_alter_table_with_event_trigger(compressed_table_id, NULL, list_make1(&cmd), true);
+}
+
+static void
+set_statistics_on_compressed_chunk(Oid compressed_table_id)
+{
+	Relation table_rel = table_open(compressed_table_id, ShareUpdateExclusiveLock);
+	Relation attrelation = table_open(AttributeRelationId, RowExclusiveLock);
+	TupleDesc table_desc = RelationGetDescr(table_rel);
+	Oid compressed_data_type = ts_custom_type_cache_get(CUSTOM_TYPE_COMPRESSED_DATA)->type_oid;
+	for (int i = 0; i < table_desc->natts; i++)
+	{
+		Form_pg_attribute attrtuple;
+		HeapTuple tuple;
+		Form_pg_attribute col_attr = TupleDescAttr(table_desc, i);
+
+		/* skip system columns */
+		if (col_attr->attnum <= 0)
+			continue;
```

Comment:
Can we have system columns in a physical tuple of relation? I think no, they can only appear in scan tuples.

Reply:
Hmm you might be right that those are skipped already but pg_attribute does include the system columns and we are dealing with Form_pg_attribute here